{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d4523923",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14705 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14707 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14710 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14713 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14717 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14719 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14723 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14729 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14732 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14737 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14743 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14745 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14754 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14761 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14764 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14766 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14769 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14771 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14773 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14775 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14778 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14780 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14783 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14785 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14787 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14789 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14791 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14793 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14795 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14797 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14799 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14801 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14803 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14805 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14807 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14809 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14812 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14814 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14816 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14818 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14820 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14823 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14825 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14827 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14832 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14834 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14836 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14838 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14840 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14842 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14845 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14847 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14849 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14851 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14853 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14855 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14857 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14859 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14861 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14863 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n",
      "C:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:154: UserWarning: Could not connect to TraCI server using port 14865 (TraCI server already finished). Retrying with different port.\n",
      "  warnings.warn((\"Could not connect to TraCI server using port %s (%s).\" +\n"
     ]
    },
    {
     "ename": "FatalTraCIError",
     "evalue": "Could not connect.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFatalTraCIError\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[3], line 8\u001b[0m\n\u001b[0;32m      5\u001b[0m THRESHOLD_CARS \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m10\u001b[39m\n\u001b[0;32m      7\u001b[0m sumoCmd \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msumo-gui\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-c\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mregion.sumocfg\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m--remote-port\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m8813\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m----> 8\u001b[0m traci\u001b[38;5;241m.\u001b[39mstart(sumoCmd)\n\u001b[0;32m      9\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAgent connecté à SUMO.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m     11\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[1;32mC:\\Program Files (x86)\\Eclipse\\Sumo\\tools\\traci\\main.py:157\u001b[0m, in \u001b[0;36mstart\u001b[1;34m(cmd, port, numRetries, label, verbose, traceFile, traceGetters, stdout, doSwitch)\u001b[0m\n\u001b[0;32m    154\u001b[0m         warnings\u001b[38;5;241m.\u001b[39mwarn((\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCould not connect to TraCI server using port \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m (\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m).\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m\n\u001b[0;32m    155\u001b[0m                        \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m Retrying with different port.\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;241m%\u001b[39m (sumoPort, e))\n\u001b[0;32m    156\u001b[0m         numRetries \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m--> 157\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m FatalTraCIError(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCould not connect.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[1;31mFatalTraCIError\u001b[0m: Could not connect."
     ]
    }
   ],
   "source": [
    "import traci\n",
    "import time\n",
    "\n",
    "BUS_TYPE = \"bus\"\n",
    "THRESHOLD_CARS = 10\n",
    "\n",
    "sumoCmd = [\"sumo-gui\", \"-c\", \"region.sumocfg\", \"--remote-port\", \"8813\"]\n",
    "traci.start(sumoCmd)\n",
    "print(\"Agent connecté à SUMO.\")\n",
    "\n",
    "try:\n",
    "    for step in range(500):\n",
    "        traci.simulationStep()\n",
    "        veh_count = traci.edge.getLastStepVehicleNumber(\"edge1\")\n",
    "        veh_list = traci.edge.getLastStepVehicleIDs(\"edge1\")\n",
    "        bus_present = any(traci.vehicle.getTypeID(v) == BUS_TYPE for v in veh_list)\n",
    "\n",
    "        traffic_light_id = \"n2\"\n",
    "        current_phase = traci.trafficlight.getPhase(traffic_light_id)\n",
    "\n",
    "        if bus_present and current_phase != 0:\n",
    "            print(\"🚌 Bus détecté, passage au vert.\")\n",
    "            traci.trafficlight.setPhase(traffic_light_id, 0)\n",
    "        elif veh_count > THRESHOLD_CARS and current_phase != 0:\n",
    "            print(f\"🚗 {veh_count} voitures détectées, passage au vert.\")\n",
    "            traci.trafficlight.setPhase(traffic_light_id, 0)\n",
    "\n",
    "        time.sleep(0.05)\n",
    "finally:\n",
    "    traci.close()\n",
    "    print(\"✅ Simulation terminée.\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
